### PR TITLE
Judges can talk in games

### LINKED ISSUE #4078: Judge role can't override no spectator chat
2.7.5

Judges should have the power to override the no spectator chat, while in judge mode. Similar to moderator/administrator access.

### PR DESCRIPTION
**Pull request by ZeldaZach**
Server & Client sided change. Will require a version bump to deploy officially.

Yes, the if statements are ugly. I left a comment block to indicate what's actually going on.

## Related Ticket(s)
- Fixes #4078

### CODE REVIEW

File: cockatrice/src/tab_game.h [hunk 1]
```diff
@@ -117,7 +117,7 @@ class TabGame : public Tab
 {
     Q_OBJECT
 private:
-    QTimer *gameTimer;
+    QTimer *gameTimer{};
```
Comment (ebbit1q) [inline cockatrice/src/tab_game.h:120]: why add default initializers to all these?

Comment (ZeldaZach) [reply cockatrice/src/tab_game.h:120]: Clangify did that, as it said some variables were missed. I'll revert this

File: cockatrice/src/tab_game.cpp [hunk 2]
```diff
@@ -1814,10 +1819,16 @@ void TabGame::createMessageDock(bool bReplay)
         sayEdit->setCompleter(completer);
         actCompleterChanged();
 
-        if (spectator && !gameInfo.spectators_can_chat() && tabSupervisor->getAdminLocked()) {
+        /* Spectators can only talk if:
+         * (a) the game creator allows it
+         * (b) the spectator is a moderator/administrator
+         * (c) the spectator is a judge
+         */
+        if (!(isSpectator() && (gameInfo.spectators_can_chat() || !tabSupervisor->getAdminLocked() || isJudge()))) {
```
Comment (ebbit1q) [inline cockatrice/src/tab_game.cpp:1827]: is there a specific reason to use (and add) these public functions while they're simply available as private members of this object?

Comment (ZeldaZach) [reply cockatrice/src/tab_game.cpp:1827]: Not really, I thought it was a bit cleaner. I can change that if you'd like

File: cockatrice/src/tab_game.cpp [hunk 3]
```diff
@@ -1814,10 +1814,16 @@ void TabGame::createMessageDock(bool bReplay)
         sayEdit->setCompleter(completer);
         actCompleterChanged();
 
-        if (spectator && !gameInfo.spectators_can_chat() && tabSupervisor->getAdminLocked()) {
+        /* Spectators can only talk if:
+         * (a) the game creator allows it
+         * (b) the spectator is a moderator/administrator
+         * (c) the spectator is a judge
+         */
+        if (!(spectator && (gameInfo.spectators_can_chat() || !tabSupervisor->getAdminLocked() || judge))) {
```
Comment (ebbit1q) [inline cockatrice/src/tab_game.cpp:1822]: I think the style this was in previously was a lot more readable, simply change this to && !judge?

Comment (ZeldaZach) [reply cockatrice/src/tab_game.cpp:1822]: That won't work, because you need to be a spectator and one of the three options. This was the cleanest way I could come up with.

`if NOT (A & (B OR (NOT C) OR D))`

Comment (ebbit1q) [reply cockatrice/src/tab_game.cpp:1822]: I got confused so I made a truth table, your code will hide the chat for everyone who isn't a spectator :(
`(a, b, c, d, not (a and (b or not c or d)), a and not b and (c or not d)) for a, b, c, d in ((bool(i & (1 << n)) for n in range(4)) for i in range(1 << 4))`

File: cockatrice/src/tab_game.cpp [hunk 4]
```diff
@@ -1814,10 +1814,26 @@ void TabGame::createMessageDock(bool bReplay)
...
+            if (gameInfo.spectators_can_chat() || !tabSupervisor->getAdminLocked() || judge) {
+                canChat = true;
+            }
+        }
+
+        if (!canChat) {
             sayLabel->hide();
             sayEdit->hide();
         }
+
```
Comment (kodabb) [inline cockatrice/src/tab_game.cpp:1836]: nit: unrelated cosmetic change

**Review by kodabb — APPROVED** [review]

File: cockatrice/src/tab_game.cpp (hunk 4 above)
Comment (ebbit1q) [reply cockatrice/src/tab_game.cpp:1836]: looks a lot better though

File: cockatrice/src/tab_supervisor.cpp [hunk 5]
```diff
@@ -621,11 +621,17 @@ void TabSupervisor::updateCurrent(int index)
...
 
+/**
+ * Determine if a user is a moderator/administrator
+ * By seeing if they have the admin or log tabs open
+ * @return if one of the two tabs are open
+ */
 bool TabSupervisor::getAdminLocked() const
 {
-    if (!tabAdmin)
+    if (!tabAdmin && !tabLog)
```
Comment (ebbit1q) [inline cockatrice/src/tab_supervisor.cpp:631]: this segfaults if tabadmin is uninitialized while tabLog is true

Comment (ZeldaZach) [reply cockatrice/src/tab_supervisor.cpp:631]: True. I'm kinda planning ahead for the other PR with this change, so I'll clean it up

File: cockatrice/src/tab_supervisor.cpp [hunk 6]
```diff
@@ -621,6 +621,11 @@ void TabSupervisor::updateCurrent(int index)
         emit setMenu();
 }
 
+/**
+ * Determine if a user is a moderator/administrator
+ * By seeing if they have the admin tab open & unlocked
+ * @return if one of the two tabs are open
```
Comment (ebbit1q) [inline cockatrice/src/tab_supervisor.cpp:627]: this isn't valid anymore

**Review by ebbit1q — APPROVED** [review]
bikeshedder seal of approval(tm)(r)(c)

**Comment by tooomm** [comment]
Does it make sense to include it into the featureset file?

**Comment by ZeldaZach** [comment]
Not really, because the server owner just has to update. It's not a breaking change on the client side.
